File: CSharp/12-copilot-skill/CustomerService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace Demo.Services;

public record Customer(int Id, string Name, string Email, string? Phone, DateTime CreatedAt);

public record CreateCustomerRequest(string Name, string Email, string? Phone = null);

public record UpdateCustomerRequest(string? Name = null, string? Email = null, string? Phone = null);

public class CustomerService
{
    private readonly Dictionary<int, Customer> _store = new();
    private int _nextId = 1;

    public async Task<Customer> CreateAsync(
        CreateCustomerRequest request,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(request);
        if (string.IsNullOrWhiteSpace(request.Name))
            throw new ArgumentException("Name cannot be empty.", nameof(request));
        if (string.IsNullOrWhiteSpace(request.Email) || !request.Email.Contains('@'))
            throw new ArgumentException("A valid email address is required.", nameof(request));

        await Task.Delay(0, cancellationToken);

        var customer = new Customer(
            _nextId++,
            request.Name.Trim(),
            request.Email.Trim().ToLowerInvariant(),
            request.Phone?.Trim(),
            DateTime.UtcNow);

        _store[customer.Id] = customer;
        return customer;
    }

    public async Task<Customer?> GetByIdAsync(
        int customerId,
        CancellationToken cancellationToken = default)
    {
        if (customerId <= 0)
            throw new ArgumentOutOfRangeException(nameof(customerId), "Customer ID must be a positive integer.");

        await Task.Delay(0, cancellationToken);
        _store.TryGetValue(customerId, out var customer);
        return customer;
    }

    public async Task<IReadOnlyList<Customer>> SearchAsync(
        string query,
        int page = 1,
        int pageSize = 20,
        CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(query))
            throw new ArgumentException("Search query cannot be empty.", nameof(query));
        if (page < 1)
            throw new ArgumentOutOfRangeException(nameof(page), "Page number must be at least 1.");
        if (pageSize is < 1 or > 100)
            throw new ArgumentOutOfRangeException(nameof(pageSize), "Page size must be between 1 and 100.");

        await Task.Delay(0, cancellationToken);

        return _store.Values
            .Where(c => c.Name.Contains(query, StringComparison.OrdinalIgnoreCase)
                     || c.Email.Contains(query, StringComparison.OrdinalIgnoreCase))
            .OrderBy(c => c.Name)
            .Skip((page - 1) * pageSize)
            .Take(pageSize)
            .ToList()
            .AsReadOnly();
    }

    public async Task<Customer> UpdateAsync(
        int customerId,
        UpdateCustomerRequest request,
        CancellationToken cancellationToken = default)
    {
        if (customerId <= 0)
            throw new ArgumentOutOfRangeException(nameof(customerId));
        ArgumentNullException.ThrowIfNull(request);

        await Task.Delay(0, cancellationToken);

        if (!_store.TryGetValue(customerId, out var existing))
            throw new KeyNotFoundException($"No customer with ID {customerId} was found.");

        var updated = existing with
        {
            Name  = string.IsNullOrWhiteSpace(request.Name)  ? existing.Name  : request.Name.Trim(),
            Email = string.IsNullOrWhiteSpace(request.Email) ? existing.Email : request.Email.Trim().ToLowerInvariant(),
            Phone = request.Phone is null ? existing.Phone : request.Phone.Trim()
        };

        _store[customerId] = updated;
        return updated;
    }

    public async Task<bool> DeleteAsync(int customerId, CancellationToken cancellationToken = default)
    {
        if (customerId <= 0)
            throw new ArgumentOutOfRangeException(nameof(customerId));

        await Task.Delay(0, cancellationToken);
        return _store.Remove(customerId);
    }

    public int Count => _store.Count;
}

File: Javascript/12-copilot-skill/customerService.js
'use strict';

class Customer {
  constructor(id, name, email, phone, createdAt) {
    this.id = id;
    this.name = name;
    this.email = email;
    this.phone = phone ?? null;
    this.createdAt = createdAt;
  }
}

class CustomerService {
  #store = new Map();
  #nextId = 1;

  async create(request) {
    if (!request?.name?.trim()) {
      throw new TypeError('Name is required.');
    }
    if (!request?.email?.includes('@')) {
      throw new TypeError('A valid email address is required.');
    }

    const customer = new Customer(
      this.#nextId++,
      request.name.trim(),
      request.email.trim().toLowerCase(),
      request.phone?.trim() ?? null,
      new Date()
    );

    this.#store.set(customer.id, customer);
    return customer;
  }

  async getById(customerId) {
    if (!Number.isInteger(customerId) || customerId <= 0) {
      throw new RangeError('customerId must be a positive integer.');
    }
    return this.#store.get(customerId) ?? null;
  }

  async search(query, page = 1, pageSize = 20) {
    if (!query?.trim()) {
      throw new TypeError('Search query cannot be empty.');
    }
    if (!Number.isInteger(page) || page < 1) {
      throw new RangeError('page must be a positive integer.');
    }
    if (!Number.isInteger(pageSize) || pageSize < 1 || pageSize > 100) {
      throw new RangeError('pageSize must be between 1 and 100.');
    }

    const lower = query.toLowerCase();
    const matches = [...this.#store.values()]
      .filter(c => c.name.toLowerCase().includes(lower) || c.email.toLowerCase().includes(lower))
      .sort((a, b) => a.name.localeCompare(b.name));

    return matches.slice((page - 1) * pageSize, page * pageSize);
  }

  async update(customerId, request) {
    if (!Number.isInteger(customerId) || customerId <= 0) {
      throw new RangeError('customerId must be a positive integer.');
    }
    const existing = this.#store.get(customerId);
    if (!existing) {
      throw new Error(`No customer with ID ${customerId} was found.`);
    }

    const updated = new Customer(
      existing.id,
      request.name?.trim() ?? existing.name,
      request.email?.trim().toLowerCase() ?? existing.email,
      request.phone !== undefined ? (request.phone?.trim() ?? null) : existing.phone,
      existing.createdAt
    );

    this.#store.set(customerId, updated);
    return updated;
  }

  async delete(customerId) {
    if (!Number.isInteger(customerId) || customerId <= 0) {
      throw new RangeError('customerId must be a positive integer.');
    }
    return this.#store.delete(customerId);
  }

  get count() {
    return this.#store.size;
  }
}

module.exports = { Customer, CustomerService };
